python/ray/experimental/channel/cpu_nccl_group.py

```diff
@@ -0,0 +1,273 @@
+                self.collective_data[op_id] = data
+                self.condition.notify_all()
+            else:
+                await self.condition.wait_for(
+                    lambda: self.num_actors_seen[op_id] == self.num_actors
+                )
+
+            data = self.collective_data[op_id]
+            self.num_actors_read[op_id] += 1
+
+            if self.num_actors_read[op_id] == self.num_actors:
+                del self.collective_data[op_id]
+                del self.num_actors_seen[op_id]
+                del self.num_actors_read[op_id]
+
+            return data
+
+    def _apply_op(self, op: ReduceOp, tensors: List[torch.Tensor]) -> torch.Tensor:
+        """Apply the specified reduction operation across a list of tensors."""
+        result = tensors[0].clone()
+        if op == ReduceOp.SUM:
+            for tensor in tensors[1:]:
+                result += tensor
+        elif op == ReduceOp.PRODUCT:
+            for tensor in tensors[1:]:
+                result *= tensor
+        elif op == ReduceOp.MAX:
+            for tensor in tensors[1:]:
+                result = torch.max(result, tensor)
+        elif op == ReduceOp.MIN:
+            for tensor in tensors[1:]:
+                result = torch.min(result, tensor)
+        elif op == ReduceOp.AVG:
+            result = sum(tensors) / len(tensors)
+        else:
+            raise ValueError(f"Operation {op} not supported")
+        return result
+
+
+class CPUNcclGroup(GPUCommunicator):
+    """
+    Mock the internal _NcclGroup to use a communicator actor instead of a NCCL group
+    for communication.
+    """
+
+    def __init__(self, world_size: int, actor_handles: List["ray.actor.ActorHandle"]):
+        """We use the op index to synchronize the sender and receiver at the
+        communicator actor."""
+        self._world_size = world_size
+        self._actor_handles = actor_handles
+        self.num_ops = defaultdict(int)
+        self.communicators = set()
+        self._rank = None
+
+    def send(self, tensor: torch.Tensor, peer_rank: int):
+        """Send the tensor to the communicator actor."""
+        comm_key = f"communicator-p2p-{self.get_self_rank()}-{peer_rank}"
+        comm = CPUCommunicator.options(name=comm_key, get_if_exists=True).remote(2)
+        self.communicators.add(comm)
+
+        ray.get(comm.wait_p2p.remote(self.num_ops[comm_key], tensor))
+        self.num_ops[comm_key] += 1
+
+    def recv(
+        self,
+        shape: Tuple[int],
+        dtype: torch.dtype,
+        peer_rank: int,
+        allocator: Optional[TorchTensorAllocator] = None,
+    ):
+        """Receive the tensor from the communicator actor."""
+        comm_key = f"communicator-p2p-{peer_rank}-{self.get_self_rank()}"
+        comm = CPUCommunicator.options(name=comm_key, get_if_exists=True).remote(2)
+        self.communicators.add(comm)
+
+        received_tensor = ray.get(comm.wait_p2p.remote(self.num_ops[comm_key]))
+        assert (
```

Comment:
In this case you can probably just directly return the `received_tensor` (`allocator` is needed for cases where a receive buffer needs to be allocated before the recv happens).